provider.go

```diff
@@ -0,0 +1,154 @@
+package goose
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"time"
+)
+
+// Provider is a goose migration provider.
+type Provider struct {
+	db  *sql.DB
+	opt *ProviderOptions
+}
+
+// NewProvider returns a new goose Provider.
+//
+// The caller is responsible for matching the database dialect with the database/sql driver. For
+// example, if the database dialect is "postgres", the database/sql driver could be
+// github.com/lib/pq or github.com/jackc/pgx.
+//
+// If opts is nil, the default options are used. See ProviderOptions for more information.
+//
+// Unless otherwise specified, all methods on Provider are safe for concurrent use.
+func NewProvider(dialect Dialect, db *sql.DB, opts *ProviderOptions) (*Provider, error) {
```

Comment:
@VojtechVitek wdyt about adding the Provider to the /v3 module instead of cutting an entire new major version /v4. 

This was the plan for https://github.com/pressly/goose/pull/507, but after a lot of thought, I think we can iterate on the Provider and extend it to support a bunch of features folks have been asking for.

Reply:
sounds like a good idea :)